### PR TITLE
Feature/subseasonal fixes part2

### PR DESCRIPTION
**Pull request by ShannonShields-NOAA**
<b>Note to developers: You must use this PR template!</b>

## Description of Changes

> Please include a summary of the changes and the related GitHub issue(s). Please also include relevant motivation and context.
This PR includes the use of split_by_subset to trim prepbufr NAM files to only include temp obs and adds checks to ensure dcom files are not corrupt. These items are listed on the Fixes/Additions doc.
## Developer Questions and Checklist
* Is this a high priorty PR? If so, why and is there a date it needs to be merged by?
* No
* Do you have any planned upcoming annual leave/PTO?
* I might need to take some time off this Friday (10/18) for my pet's surgery, but it won't be the full day.
* Are there any changes needed for when the jobs are supposed to run?
  No
- [ ] The code changes follow [NCO's EE2 Standards](https://www.nco.ncep.noaa.gov/idsb/implementation_standards/ImplementationStandards.v11.0.0.pdf).
- [ ] Developer's name is removed throughout the code and have used `${USER}` where necessary throughout the code.
- [ ] References the feature branch for `HOMEevs` are removed from the code.
- [ ] J-Job environment variables, COMIN and COMOUT directories, and output follow what has been [defined](https://docs.google.com/document/d/1JWg_4q80aYmmAoD21GFjp9R9y5-3w7WGM3-0HJk0Pjs/edit#heading=h.7ysbr191vzu4) for EVS.
- [ ] Jobs over 15 minutes in runtime have restart capability.
- [ ] If applicable, changes in the `dev/drivers/scripts` or `dev/modulefiles` have been made in the corresponding `ecf/scripts` and `ecf/defs/evs-nco.def`? 
- [ ] Jobs contain the approriate file checking and don't run METplus for any missing data.
- [ ] Code is using METplus wrappers structure and not calling MET executables directly.
- [ ] Log is free of any ERRORs or WARNINGs.

## Testing Instructions

> Please include testing instructions for the PR assignee. Include all relevant input datasets needed to run the tests.
Checkout this feature branch and link FIXevs directory.
Set COMOUT to respective assignee path for prep driver.
Please run the following script for testing the described changes:
dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.sh
***NOTE**: I do not know how to test for corrupted dcom files (I only know how to test missing), so the file checks for this will need to be tested in this PR. I followed Mallory's code for this so she and Perry likely know what needs to be done for proper testing.
Also, I noticed a discrepancy in the chmod number used between global_det and subseasonal components. I have "run_shell_command(['chmod', '**640**', dest_file])" while 
global_det has "run_shell_command(['chmod', '**750**', dest_file])". Which number should be used for restricted data like ECMWF and prepbufr NAM?

### CODE REVIEW

**Comment by malloryprow** [comment]
Can you add the module load for bufr into the appropriate prep ecf script(s)?

File: ush/subseasonal/subseasonal_prep_obs.py [hunk 1]
```diff
@@ -184,19 +186,18 @@
...
-                        )
-                if not os.path.exists(prod_file):
-                    sub_util.log_missing_file_obs(
-                        log_missing_file, prod_file, OBS,
-                        CDATE_dt
+                    sub_util.prep_prod_prepbufr_file(
+                        prod_file, arch_file, CDATE_dt,
+                        log_missing_file
                     )
+                    #sub_util.copy_file(prod_file, arch_file)
```
Comment (malloryprow) [inline ush/subseasonal/subseasonal_prep_obs.py]: Can you remove this commented out code if it is not needed?

**Comment by malloryprow** [comment]
chmod 640 - gives read and write permissions to the owner, read permissions to the group, and no permissions to public
chmod 750 - gives read, write, and execute permissions to the owner, read and execute permissions to the group, and no permissions to the publlic

I don't think either is wrong; they both accomplish the goal of people outside the rstprod group not having any permissions on the file.

**Comment by ShannonShields-NOAA** [comment]
@malloryprow In dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.sh I have: export OBSNAME="gfs ecmwf osi ghrsst umd nam ccpa"
However, in ecf/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.ecf I have: export OBSNAME="gfs ecmwf osi ghrsst umd nam"
Should these scripts be matching (should I add "ccpa" to ecf)?
& I see that I also need to add "module load met/${met_ver}" and "module load metplus/${metplus_ver}" to prep ecf script.

**Comment by malloryprow** [comment]
> @malloryprow In dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.sh I have: export OBSNAME="gfs ecmwf osi ghrsst umd nam ccpa" However, in ecf/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.ecf I have: export OBSNAME="gfs ecmwf osi ghrsst umd nam" Should these scripts be matching (should I add "ccpa" to ecf)? & I see that I also need to add "module load met/${met_ver}" and "module load metplus/${metplus_ver}" to prep ecf script.

If you want that eventually a part of prep when it runs on operations, I would include it.

**Comment by malloryprow** [comment]
I'll kick off the job at 1400Z, which is when the job is set to run in the parallels.

It would be good to do a test after that with corrupted data. What I did for this was I created my own psuedo com/dcom that was filled with bad data. What I did to make the corrupted data was I would copy it from prod and then interrupt the transfer so the file wasn't complete. 

`rsync -ahr --inplace --progress --bwlimit=100 dlogin01:/lfs/h1/ops/prod/dcom/20240529/wgrbbul/ecmwf/U1D0529* .`

That is an example of what I used to do the transfer. If you need help building the corrupted data, shoot me a message in GChat.

**Comment by ShannonShields-NOAA** [comment]
@malloryprow I think I managed to corrupt the ECMWF file...I made my own dcom directory under /lfs/h2/emc/vpppg/noscrub/shannon.shields/dcom/20241015/wgrbbul/ecmwf. I see the full file should be:
-rw-r----- 1 shannon.shields rstprod **6.9M** Oct 15 05:40 DCD10150000101500001
When I interrupted the transfer process, the file looks like:
-rw------- 1 shannon.shields vpppg **1.3M** Oct 17 13:27 DCD10150000101500001
Does this seem corrupted to you? (I still need to test it with my new code changes, but thought I would get input on this first.)

**Comment by malloryprow** [comment]
Seems to me it should be good.

**Comment by malloryprow** [comment]
### jevs_subseasonal_obs_prep.sh - Clean Run
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr590/EVS/dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.o158484074
**COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr590/evs/v2.0/prep/subseasonal/atmos
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_subseasonal_obs_prep.158484074.cbqs01

**Comment by ShannonShields-NOAA** [comment]
@malloryprow Everything looks good for the obs prep job; I see the split_by subset command working on the prepbufr nam files.
I do see this message in the output log file:
 ++++++++++++++BUFR ARCHIVE LIBRARY+++++++++++++++++
 BUFRLIB: MAXOUT - THE RECORD LENGTH OF ALL BUFR MESSAGES CREATED FROM THIS POINT ON IS BEING CHANGED FROM   10000 TO   20000
Since this is the first time using bufr module, I am not sure what this is exactly indicating..is this ok?

**Comment by malloryprow** [comment]
I think so. It is what I have seen from all other usage of split_by_subset.

**Comment by ShannonShields-NOAA** [comment]
Ok, so for the corrupt dcom file test, please make sure to remove the ecmwf, ghrsst_ospo, osi_saf, and umd directories from /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr590/evs/v2.0/prep/subseasonal/atmos.20241015/. The scripts check if these directories exist, and if they do, the jobs won't rerun.
Then in dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.sh, set the following to point to the psuedo dcom path:
export DCOMINecmwf=/lfs/h2/emc/vpppg/noscrub/shannon.shields/dcom
export DCOMINghrsst=/lfs/h2/emc/vpppg/noscrub/shannon.shields/dcom
export DCOMINosi=/lfs/h2/emc/vpppg/noscrub/shannon.shields/dcom
export DCOMINumd=/lfs/h2/emc/vpppg/noscrub/shannon.shields/dcom
I have corrupted all of these files, so there should be corrupt WARNINGs for all.
Then rerun jevs_subseasonal_obs_prep.sh.

**Comment by malloryprow** [comment]
### jevs_subseasonal_obs_prep.sh - Corrupt Run
**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr590/EVS/dev/drivers/scripts/prep/subseasonal/jevs_subseasonal_obs_prep.o158490016
**COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr590/evs/v2.0/prep/subseasonal/atmos
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_subseasonal_obs_prep.158490016.cbqs01

**Comment by ShannonShields-NOAA** [comment]
jevs_subseasonal_obs_prep.sh - Corrupt Run job ran as expected and reported WARNINGs for each of the corrupt dcom files (and did not generate the final output prep files for the associated directories). All else ran fine. If everything looks good on your end (I just checked that this branch is up-to-date), then this PR should be ready for merging. Thank you @malloryprow for your help, especially for the corrupt dcom testing.

**Review by malloryprow — APPROVED** [review]
Code changes are good and testing successful.

**Review by AndrewBenjamin-NOAA — APPROVED** [review]
I have reviewed the code and accept these changes. Also, thank you for working with Mallory on the additional corrupt data testing.

**Comment by malloryprow** [comment]
Thanks @ShannonShields-NOAA! If this addresses anything in the Fixes and Additions document, be sure to ✔️ it off!

**Pull request merged**